diff --git c/base/runtime/include/ml-unixdep.h i/base/runtime/include/ml-unixdep.h
index 23989df..6097ad3 100644
--- c/base/runtime/include/ml-unixdep.h
+++ i/base/runtime/include/ml-unixdep.h
@@ -279,6 +279,8 @@ extern char     *sys_errlist[];
 #  define HAS_PARTIAL_MUNMAP
 #  define HAS_SELECT
 #  define HAS_STRERROR
+#  define HAS_MKSTEMP
+#  define __USE_GNU
 
 #include <features.h>
 
diff --git c/base/runtime/objs/mk.ppc-linux i/base/runtime/objs/mk.ppc-linux
index adaff82..ef10c37 100644
--- c/base/runtime/objs/mk.ppc-linux
+++ i/base/runtime/objs/mk.ppc-linux
@@ -16,7 +16,7 @@ XLIBS =
 LD_LIBS =
 XDEFS =
 BASE_DEFS =
-DEFS		= $(XDEFS) $(BASE_DEFS) -DHOST_PPC -DTARGET_PPC -DOPSYS_UNIX -DOPSYS_LINUX -D_POSIX_SOURCE -D_BSD_SOURCE -DGNU_ASSEMBLER
+DEFS		= $(XDEFS) $(BASE_DEFS) -DHOST_PPC -DTARGET_PPC -DOPSYS_UNIX -DOPSYS_LINUX -D_GNU_SOURCE -D_POSIX_SOURCE -D_BSD_SOURCE -DGNU_ASSEMBLER
 TARGET =	PPC
 VERSION =	v-ppc-linux
 RUNTIME =	run.ppc-linux
